test/test_transferagent.py

```diff
@@ -40,6 +41,31 @@ def store_file_object(self, key, fd, *, cache_control=None, metadata=None, mimet
         raise StorageError("foo")
 
 
+class MockStorageNetworkThrottle(MockStorage):
+    """
+    Storage simulating network throttling when uploading file chunks to object storage.
+    """
+    NUM_CHUNKS = 4
+    INCREMENT_WAIT_PER_CHUNK = [0.1, 0.1, 1, 0.1]
+
+    # pylint: disable=unused-argument
+    def store_file_object(self, key, fd, *, cache_control=None, metadata=None, mimetype=None, upload_progress_fn=None):
+        file_size = int(metadata["Content-Length"]) if "Content-Length" in metadata else None
+        if not file_size:
+            return
+
+        chunk_size = round(file_size / self.NUM_CHUNKS)
+        for chunk_num in range(self.NUM_CHUNKS):
+            time.sleep(self.INCREMENT_WAIT_PER_CHUNK[chunk_num])
+            if upload_progress_fn:
+                upload_progress_fn(chunk_size)
+
+
+class PatchedUploadEventProgressTracker(UploadEventProgressTracker):
+    CHECK_FREQUENCY = .2
+    WARNING_TIMEOUT = .5
```

Comment:
Doesn't that introduce some flakyness ? I'm sorry it's not very constructive, as I don't have much to back it up but that means we need to check 2 out of 5 possible times in the alloted 1s. Maybe we could reduce both the check frequency and warning timeout, or raise the wait for the third chunk ?

Reply:
Not sure if you commented on the right lines.   I could increment the wait time for the 3rd chunk, I left it to 1 sec since I don't want the test to take that long.  But can make it to 1.5 :) whole point is to exceed the average rate and warning timeout.  Check frequency I think its okayish
https://github.com/Aiven-Open/pghoard/blob/7efc655640e6ce0a90ce588f484f134f494dce0f/test/test_transferagent.py#L44-L49

Reply:
Btw, I am not fan of  `PatchedUploadEventProgressTracker` and `MockStorageNetworkThrottle`, but the way how the tests are setup  makes it difficult to use `@patch.object` since PGHoardTestCase already sets up all pghoard object  :crying_cat_face:

Reply:
discussed through slack, will keep as it is